crates/control/src/role_assignment.rs

```diff
@@ -372,289 +285,320 @@ impl RoleAssignment {
+}
+
+fn is_cooldown_elapsed(now: SystemTime, last: Option<SystemTime>, cooldown: Duration) -> bool {
+    match last {
+        None => true,
+        Some(last_time) => now.duration_since(last_time).expect("time ran backwards") > cooldown,
+    }
+}
+
+fn is_enough_message_budget_left(context: &CycleContext<impl NetworkInterface>) -> bool {
+    context
+        .filtered_game_controller_state
+        .is_some_and(|game_controller_state| {
+            game_controller_state.remaining_number_of_messages
+                > context
+                    .spl_network
+                    .remaining_amount_of_messages_to_stop_sending
+        })
+}
+
+#[derive(Clone, Copy, Debug)]
+enum Event {
+    None,
+    Striker(StrikerEvent),
+    Loser,
+}
+
+#[derive(Clone, Copy, Debug)]
+struct StrikerEvent {
+    player_number: PlayerNumber,
+    time_to_reach_kick_position: Duration,
 }
 
 #[allow(clippy::too_many_arguments)]
 fn process_role_state_machine(
     current_role: Role,
-    current_pose: Isometry2<Ground, Field>,
-    detected_own_ball: Option<&BallPosition<Ground>>,
+    detected_own_ball: bool,
     primary_state: PrimaryState,
-    incoming_message: Option<&StrikerMessage>,
+    event: Event,
     time_to_reach_kick_position: Option<Duration>,
-    send_spl_striker_message: bool,
     team_ball: Option<BallPosition<Field>>,
     cycle_start_time: SystemTime,
     filtered_game_controller_state: Option<&FilteredGameControllerState>,
     player_number: PlayerNumber,
-    striker_trusts_team_ball: Duration,
+    striker_trusts_team_ball_duration: Duration,
     optional_roles: &[Role],
-) -> (Role, bool, Option<BallPosition<Field>>) {
+) -> Role {
     if let Some(game_controller_state) = filtered_game_controller_state {
-        match game_controller_state.game_phase {
-            GamePhase::PenaltyShootout {
-                kicking_team: Team::Hulks,
-            } => return (Role::Striker, false, None),
-            GamePhase::PenaltyShootout {
-                kicking_team: Team::Opponent,
-            } => return (Role::Keeper, false, None),
-            _ => {}
+        if let GamePhase::PenaltyShootout { kicking_team } = game_controller_state.game_phase {
+            return match kicking_team {
+                Team::Hulks => Role::Striker,
+                Team::Opponent => Role::Keeper,
+            };
         };
         if let Some(SubState::PenaltyKick) = game_controller_state.sub_state {
-            return (current_role, false, None);
+            return current_role;
         }
     }
 
     if primary_state != PrimaryState::Playing {
-        match detected_own_ball {
-            None => return (current_role, false, team_ball),
-            Some(..) => {
-                return (
-                    current_role,
-                    false,
-                    team_ball_from_seen_ball(detected_own_ball, current_pose, cycle_start_time),
-                )
-            }
-        }
+        return current_role;
     }
```

Comment:
why do we trigger the state machine during all the other states? i'd expected role assignment to return early in the current way we handle ready and set...

Reply:
Moved out of the state machine